includes/forms/class-llms-forms-dynamic-fields.php

```diff
@@ -209,6 +210,151 @@ public function modify_account_form( $blocks, $location ) {
 
 	}
 
+	/**
+	 * Maybe add the required email and password block to a form
+	 *
+	 * @since [version]
+	 *
+	 * @param array[] $blocks   Array of parsed WP_Block arrays.
+	 * @param string  $location The request form location ID.
+	 * @param array   $args     Additional arguments passed to the short-circuit filter.
+	 * @return array[]
+	 */
+	public function maybe_add_required_block_fields( $blocks, $location, $args ) {
+
+		$fields_to_require = array();
+
+		if ( ( ! is_user_logged_in() && in_array( $location, array( 'checkout', 'registration' ), true ) ) ||
+				( is_user_logged_in() && 'account' === $location ) ) {
+			$fields_to_require = array(
+				// Field ID => block name.
+				'email_address' => 'email',
+				'password'      => 'password',
+			);
+		}
+
+		/**
+		 * Filters the required block fields to add to the form
+		 *
+		 * @since [version]
+		 *
+		 * @param array[] $fields   Array of field_id => block_name required.
+		 * @param array[] $blocks   Array of parsed WP_Block arrays.
+		 * @param string  $location The request form location ID.
+		 * @param array   $args     Additional arguments passed to the short-circuit filter.
+		 */
+		$fields_to_require = apply_filters( 'llms_forms_required_block_fields', $fields_to_require, $blocks, $location, $args );
+
+		if ( empty( $fields_to_require ) ) {
+			return $blocks;
+		}
+
+		foreach ( $fields_to_require as $field_id => $field_block_name ) {
+
+			$block = $this->find_block( $field_id, $blocks );
+
+			if ( ! empty( $block ) ) {
+
+				$blocks = $this->make_block_visible( $block[1], $blocks, $block[0] );
+
+				unset( $fields_to_require[ $field_id ] );
+				if ( empty( $fields_to_require ) ) { // All the required blocks are present.
+					return $blocks;
+				}
+			}
+		}
+
+		$blocks_to_add = array();
+		foreach ( $fields_to_require as $field_id => $block_to_add ) {
+			$blocks_to_add[] = LLMS_Form_Templates::get_block( $block_to_add, $location, true );
```

Comment:
I think that *if* the reusable block exists it should be used here but if it *doesn't* exist it will be created on the page load (and then used again later)

And this just doens't feel like the proper way to handle this to me although it does make it more possible for users to customize the field.

The *best* scenario I think would be to lookup if the reusable exists and then use it if it does and use a non-reusable if it doesn't but 

Modified some things on my PR, thoughts?

Reply:
@thomasplevy 
All you say makes sense. And all I see in your PR makes sense too. I could just merge it now, what do you think? Then I'll give it a better review tomorrow.